### PR TITLE
Remove global lock when pruning tenants

### PR DESCRIPTION
**Pull request by fpetkovski**
We sometimes see receiver OOMs when pruning tenants. I suspect this is due to the global lock we acquire during pruning which causes ingestion for all tenants to stop for a brief amount of time.

This commit replaces the global write lock with a read lock, and acquires tenant specific locks in places where they are least disruptive.

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

### CODE REVIEW

**Comment by fpetkovski** [comment]
I will convert this to draft while I test it in an environment.

**Comment by fpetkovski** [comment]
This has been running for 24h and I haven't see any obvious issues. Pruning also works as before and `go test -race` seems to be happy.

**Comment by yeya24** [comment]
> We sometimes see receiver OOMs when pruning tenants. I suspect this is due to the global lock we acquire during pruning which causes ingestion for all tenants to stop for a brief amount of time.

As you suspected that the global lock causes the OOM kill issue, did this change help the OOM kill in your test?

Overall using the read lock there makes sense. But I am not 100% sure that if there is any other edge cases.

**Review by saswatamcode — APPROVED** [review]
LGTM mod @yeya24's question.

**Comment by fpetkovski** [comment]
I've been running PR this for about a week in our biggest cluster and haven't noticed any issues related to the new locking mechanism.

I also tested tenant eviction and pruning in the past two days and it seems that the problem with memory spikes is resolved. The global write lock was likely causing requests to hang and potentially timeout, which then leads to retries and a backlog of open appenders.

**Review by saswatamcode — APPROVED** [review]
LGTM!

File: pkg/receive/multitsdb.go [hunk 1]
```diff
@@ -356,10 +380,12 @@ func (t *MultiTSDB) pruneTSDB(ctx context.Context, logger log.Logger, tenantInst
 		return false, err
 	}
 
-	if err := os.RemoveAll(tenantTSDB.db.Dir()); err != nil {
+	if err := os.RemoveAll(tdb.Dir()); err != nil {
 		return false, err
 	}
 
+	tenantInstance.readyS.set(nil)
```
Comment (matej-g) [inline pkg/receive/multitsdb.go:387]: Could we hit an 'unlucky' panic when returning [appender](https://github.com/thanos-io/thanos/blob/main/pkg/receive/multitsdb.go#L583) e.g. in writer, when we have `nil` storage but the tenant was not yet pruned?

**Review by matej-g — COMMENTED** [review]
Looks good 👍 , just a question about a potential edge case

File: pkg/receive/multitsdb.go (hunk 1 above)
Comment (fpetkovski) [reply pkg/receive/multitsdb.go:387]: From what I see we always handle `nil` storage in the codebase. Storage will also be nil when a tenant is not yet initialized, so it doesn't seem to be a new case.

Comment (matej-g) [reply pkg/receive/multitsdb.go:387]: You're right, resolving this 👍

**Review by matej-g — APPROVED** [review]
💯
